File: testers/avro/java/generated-and-checked-in/com/example/events/precisetypes/Decimal18_4.java
package com.example.events.precisetypes;

import dev.typr.foundations.data.precise.DecimalN;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Optional;

public record Decimal18_4(BigDecimal value) implements DecimalN {
  @java.lang.Deprecated
  public Decimal18_4 {}

  public Decimal18_4 withValue(BigDecimal value) {
    return new Decimal18_4(value);
  }

  @Override
  public java.lang.String toString() {
    return value.toString();
  }

  public static Optional<Decimal18_4> of(BigDecimal value) {
    BigDecimal scaled = value.setScale(4, RoundingMode.HALF_UP);
    return scaled.precision() <= 18 ? Optional.of(new Decimal18_4(scaled)) : Optional.empty();
  }

  public static Decimal18_4 of(Integer value) {
    return new Decimal18_4(BigDecimal.valueOf((long) (value)));
  }

  public static Optional<Decimal18_4> of(Long value) {
    return Decimal18_4.of(BigDecimal.valueOf(value));
  }

  public static Optional<Decimal18_4> of(Double value) {
    return Decimal18_4.of(BigDecimal.valueOf(value));
  }

  public static Decimal18_4 unsafeForce(BigDecimal value) {
    BigDecimal scaled = value.setScale(4, RoundingMode.HALF_UP);
    if (scaled.precision() > 18) {
      throw new IllegalArgumentException("Value exceeds precision(18, 4)");
    }
    ;
    return new Decimal18_4(scaled);
  }

  @Override
  public BigDecimal decimalValue() {
    return value;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (!(obj instanceof DecimalN other)) return false;
    return decimalValue().compareTo(other.decimalValue()) == 0;
  }

  @Override
  public int hashCode() {
    return decimalValue().stripTrailingZeros().hashCode();
  }

  @Override
  public int precision() {
    return 18;
  }

  @Override
  public int scale() {
    return 4;
  }

  @Override
  public boolean semanticEquals(DecimalN other) {
    return (other == null ? false : decimalValue().compareTo(other.decimalValue()) == 0);
  }

  @Override
  public int semanticHashCode() {
    return decimalValue().stripTrailingZeros().hashCode();
  }
}

File: testers/avro/kotlin-json/generated-and-checked-in/com/example/events/precisetypes/Decimal18_4.kt
package com.example.events.precisetypes

import dev.typr.foundations.data.precise.DecimalN
import java.lang.IllegalArgumentException
import java.math.BigDecimal
import java.math.RoundingMode

@kotlin.ConsistentCopyVisibility
data class Decimal18_4 private constructor(val value: BigDecimal) : DecimalN {
  override fun decimalValue(): BigDecimal = value

  override fun equals(other: Any?): kotlin.Boolean {
    if (this === other) return true
    if (other !is DecimalN) return false
    return decimalValue().compareTo(other.decimalValue()) == 0
  }

  override fun hashCode(): Int = decimalValue().stripTrailingZeros().hashCode()

  override fun precision(): Int = 18

  override fun scale(): Int = 4

  override fun semanticEquals(other: DecimalN): kotlin.Boolean = if (other == null) false else decimalValue().compareTo(other.decimalValue()) == 0

  override fun semanticHashCode(): Int = decimalValue().stripTrailingZeros().hashCode()

  override fun toString(): kotlin.String {
    return value.toString()
  }

  companion object {
    fun of(value: BigDecimal): Decimal18_4? {
      val scaled = value.setScale(4, RoundingMode.HALF_UP)
      return if (scaled.precision() <= 18) Decimal18_4(scaled) else null
    }

    fun of(value: Int): Decimal18_4 = Decimal18_4(BigDecimal.valueOf(value.toLong()))

    fun of(value: kotlin.Long): Decimal18_4? = Decimal18_4.of(BigDecimal.valueOf(value))

    fun of(value: kotlin.Double): Decimal18_4? = Decimal18_4.of(BigDecimal.valueOf(value))

    fun unsafeForce(value: BigDecimal): Decimal18_4 {
      val scaled = value.setScale(4, RoundingMode.HALF_UP)
      if (scaled.precision() > 18) throw IllegalArgumentException("Value exceeds precision(18, 4)")
      return Decimal18_4(scaled)
    }
  }
}